codec-http2/src/main/java/io/netty/handler/codec/http2/HpackHuffmanDecoder.java

```diff
@@ -64,183 +53,4656 @@
      * @throws Http2Exception EOS Decoded
      */
     public AsciiString decode(ByteBuf buf, int length) throws Http2Exception {
-        processor.reset();
-        buf.forEachByte(buf.readerIndex(), length, processor);
-        buf.skipBytes(length);
-        return processor.end();
+        if (length == 0) {
+            return AsciiString.EMPTY_STRING;
+        }
+        byte[] dest = new byte[length * 8 / 5];
+        int k = 0;
+        int state = 0;
+        int flags = 0;
+        for (int i = 0; i < length; i++) {
+            byte input = buf.readByte();
+            int index = (state << 4) | ((input & 0xFF) >>> 4);
+            int row = HUFFS[index];
+            flags = row & 0x00FF00;
+            if ((flags & HUFFMAN_FAIL_SHIFT) != 0) {
+                throw BAD_ENCODING;
+            } else if ((flags & HUFFMAN_EMIT_SYMBOL_SHIFT) != 0) {
+                dest[k++] = (byte) (row & 0xFF);
+            }
+            state = row >> 16;
+
+            index = (state << 4) | (input & 0x0F);
+            row = HUFFS[index];
+            flags = row & 0x00FF00;
+            if ((flags & HUFFMAN_FAIL_SHIFT) != 0) {
+                throw BAD_ENCODING;
+            } else if ((flags & HUFFMAN_EMIT_SYMBOL_SHIFT) != 0) {
+                dest[k++] = (byte) (row & 0xFF);
+            }
+            state = row >> 16;
+        }
+        if ((flags & HUFFMAN_COMPLETE_SHIFT) != HUFFMAN_COMPLETE_SHIFT) {
+            throw BAD_ENCODING;
+        }
+        return new AsciiString(dest, 0, k, false);
     }
 
-    private static final class Node {
+    private static final byte HUFFMAN_COMPLETE = 1;
+    private static final byte HUFFMAN_EMIT_SYMBOL = 1 << 1;
+    private static final byte HUFFMAN_FAIL = 1 << 2;
 
-        private final int symbol;      // terminal nodes have a symbol
-        private final int bits;        // number of bits matched by the node
-        private final Node[] children; // internal nodes have children
+    private static final int HUFFMAN_COMPLETE_SHIFT = 1 << 8;
+    private static final int HUFFMAN_EMIT_SYMBOL_SHIFT = 1 << 9;
+    private static final int HUFFMAN_FAIL_SHIFT = 1 << 10;
 
-        /**
-         * Construct an internal node
-         */
-        Node() {
-            symbol = 0;
-            bits = 8;
-            children = new Node[256];
-        }
+    private static final int[] HUFFS = new int[] {
```

Comment:
A comment for meaning of each byte in the int would be nice.

Something akin to:
(new state, flags byte, emitted byte)

I recognize the reader will have to read the code to get a good understanding.

Reply:
+1 .. @carl-mastrangelo can you add something like this ?

Reply:
also can me move the declaration of all the static fields before the method. That is usually how we do it in Netty.

Reply:
Done.